connectors/sources/mssql.py

```diff
@@ -62,6 +62,244 @@ def all_schemas(self):
+                    table=table,
+                ),
+            )
+        )
+        return row_count
+
+    async def get_table_primary_key(self, table):
+        return await anext(
+            self._execute_query(
+                query=self.queries.table_primary_key(
+                    schema=self.schema,
+                    table=table,
+                ),
+            )
+        )
+
+    async def get_table_last_update_time(self, table):
+        return await anext(
+            self._execute_query(
+                query=self.queries.table_last_update_time(
+                    schema=self.schema,
+                    table=table,
+                ),
+            )
+        )
+
+    async def data_streamer(self, table):
+        """Streaming data from a table
+
+        Args:
+            table (str): Table.
+
+        Raises:
+            exception: Raise an exception after retrieving
+
+        Yields:
+            list: It will first yield the column names, then data in each row
+        """
+        async for data in self._execute_query(
+            query=self.queries.table_data(
+                schema=self.schema,
+                table=table,
+            ),
+            fetch_many=True,
+            schema=self.schema,
+            table=table,
+        ):
+            yield data
+
+    async def _execute_query(self, query, fetch_many=False, **kwargs):
+        """Executes a query and yield rows
+
+        Args:
+            query (str): Query.
+            fetch_many (bool): Flag to use fetchmany method. Defaults to False.
+
+        Raises:
+            exception: Raise an exception after retrieving
+
+        Yields:
+            list: Column names and query response
+        """
+        retry = 1
+        yield_once = True
+
+        rows_fetched = 0
+
+        while retry <= self.retry_count:
+            try:
+                cursor = await self.get_cursor(query=query)
+                if fetch_many:
+                    # sending back column names only once
+                    if yield_once:
+                        yield [
+                            f"{kwargs['schema']}_{kwargs['table']}_{column}".lower()
+                            for column in cursor.keys()  # pyright: ignore
+                        ]
+                        yield_once = False
+
+                    while True:
+                        rows = cursor.fetchmany(size=self.fetch_size)  # pyright: ignore
+                        rows_length = len(rows)
+
+                        if not rows_length:
+                            break
+
+                        for row in rows:
+                            yield row
+
+                        rows_fetched += rows_length
+                        await asyncio.sleep(0)
+                else:
+                    yield cursor.fetchall()  # pyright: ignore
+                break
+            except (InternalClientError, ProgrammingError):
+                raise
```

Comment:
Why only these two?

Reply:
These are from the original code:
https://github.com/elastic/connectors-python/blob/e9b4e577b4311d8fd8ca01cddae14ec5528ae9c7/connectors/sources/generic_database.py#L247-L248

They are also caught here:

https://github.com/elastic/connectors-python/blob/e9b4e577b4311d8fd8ca01cddae14ec5528ae9c7/connectors/sources/generic_database.py#L434-L437

I don't know excatly why

Reply:
Hi @moxarth-elastic, git blame told me it was added by you. Why these 2 exceptions are not retryable?

Reply:
hey @wangch079, `InternalClientError` is raised by asyncpg lib when it does not process the request like "invalid datatype" and same goes for `ProgrammingError`, in sqlalchemy when we pass wrong query then it raises ProgrammingError. i.e; table not found. So, I don't think we can retry for this type of exceptions.